### DISCUSSION TITLE
Stream decrypted buffer on the fly

### DISCUSSION THREAD

**Discussion by esrk** [ente-io/ente#963] [Enhancements]
Related to [this](https://roadmap.ente.io/progressive-playback-p-8153/) suggestion.
Currently [download](https://github.com/ente-io/photos-app/blob/main/lib/utils/file_download_util.dart), [decrypt](https://github.com/ente-io/photos-app/blob/main/lib/utils/crypto_util.dart#L118) and file view are sequential operations, i.e each operation runs for the full file before the next can begin. While this is relatively fine for images, viewing videos is currently not optimal, especially for large files.

An alternative would be to stream buffers in all the 3 operations in parallel threads. Ente encrypts in blocks of size 4MB, so each operation can chain the response in blocks of `decryptionChunkSize` bytes to the next one without waiting for the complete file.

Implementing cache could be relatively trickier, especially with remote seek being possible now. Maybe this could be shipped as a beta feature which disables video controls until cache is implemented for buffers. IMO even without cache, progressive video playing is a huge win.

Note: The same buffer loading could be reused for images too.

**Comment by vishnukvmd** [comment]
Hey @esrk, this is a great suggestion. Thanks for creating the issue.

From what I recall, the current video player does not accept a buffer/stream. So we might have to modify the underlying player to make this happen.

I'll keep this issue open since this is something we intend to ship.

**Comment by iksnawias** [comment]
Please implement the streaming option for videos, that would be awesome!

**Comment by brog-io** [comment]
This has been added if I understand it right

**Comment by vishnukvmd** [comment]
Streaming is now available: https://help.ente.io/photos/faq/video-streaming
